geode-core/src/main/java/org/apache/geode/internal/cache/TXState.java

```diff
@@ -898,47 +898,65 @@ void doCleanup() {
           this.proxy.getTxMgr().getCachePerfStats()
               .incTxConflictCheckTime(statisticsClock.getTime() - conflictStart);
       }
-      Iterator<Map.Entry<InternalRegion, TXRegionState>> it = this.regions.entrySet().iterator();
+      cleanupTXRegionState();
+    } finally {
+      synchronized (this.completionGuard) {
+        this.completionGuard.notifyAll();
+      }
+      if (exception != null && !this.proxy.getCache().isClosed()) {
+        throw exception;
+      }
+    }
+  }
+
+  private void cleanupTXRegionState() {
```

Comment:
Now the logic is separated, can we add unit tests for cleanupTxRegionState() and unlockPrimaryMoveReadLock() functions.